mm2src/mm2_bitcoin/spv_validation/src/work.rs

```diff
@@ -81,25 +82,30 @@ fn retarget_timespan(retarget_timestamp: u32, last_timestamp: u32) -> u32 {
 
 async fn btc_retarget_bits(
     coin: &str,
-    height: u32,
-    last_block_header: BlockHeader,
+    last_block_header: SPVBlockHeader,
     storage: &dyn BlockHeaderStorageOps,
+    retarget_header: &SPVBlockHeader,
```

Comment:
I believe `retarget_header` parameter is not needed, please consider the following changes as a consequence of removing it.
```diff
Index: mm2src/mm2_bitcoin/spv_validation/src/work.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/mm2_bitcoin/spv_validation/src/work.rs b/mm2src/mm2_bitcoin/spv_validation/src/work.rs
--- a/mm2src/mm2_bitcoin/spv_validation/src/work.rs	(revision d380e5c32a4364a98e9e0b9b173cfebc3fb5025a)
+++ b/mm2src/mm2_bitcoin/spv_validation/src/work.rs	(date 1674558847483)
@@ -51,21 +51,11 @@
     last_block_header: SPVBlockHeader,
     storage: &dyn BlockHeaderStorageOps,
     algorithm: &DifficultyAlgorithm,
-    retarget_header: &SPVBlockHeader,
 ) -> Result<BlockHeaderBits, NextBlockBitsError> {
     match algorithm {
-        DifficultyAlgorithm::BitcoinMainnet => {
-            btc_mainnet_next_block_bits(coin, last_block_header, storage, retarget_header).await
-        },
+        DifficultyAlgorithm::BitcoinMainnet => btc_mainnet_next_block_bits(coin, last_block_header, storage).await,
         DifficultyAlgorithm::BitcoinTestnet => {
-            btc_testnet_next_block_bits(
-                coin,
-                current_block_timestamp,
-                last_block_header,
-                storage,
-                retarget_header,
-            )
-            .await
+            btc_testnet_next_block_bits(coin, current_block_timestamp, last_block_header, storage).await
         },
     }
 }
@@ -84,7 +74,6 @@
     coin: &str,
     last_block_header: SPVBlockHeader,
     storage: &dyn BlockHeaderStorageOps,
-    retarget_header: &SPVBlockHeader,
 ) -> Result<BlockHeaderBits, NextBlockBitsError> {
     let max_bits_compact: Compact = MAX_BITS_BTC.into();
 
@@ -93,18 +82,14 @@
         return Ok(BlockHeaderBits::Compact(max_bits_compact));
     }
 
-    let retarget_header = if retarget_ref == retarget_header.height {
-        retarget_header.clone()
-    } else {
-        let header = storage
-            .get_block_header(retarget_ref)
-            .await?
-            .ok_or(NextBlockBitsError::NoSuchBlockHeader {
-                coin: coin.into(),
-                height: retarget_ref,
-            })?;
-        SPVBlockHeader::from_block_header(retarget_ref, header)
-    };
+    let retarget_header = storage
+        .get_block_header(retarget_ref)
+        .await?
+        .map(|h| SPVBlockHeader::from_block_header(retarget_ref, h))
+        .ok_or(NextBlockBitsError::NoSuchBlockHeader {
+            coin: coin.into(),
+            height: retarget_ref,
+        })?;
 
     // timestamp of block(height - RETARGETING_INTERVAL)
     let retarget_timestamp = retarget_header.time;
@@ -130,7 +115,6 @@
     coin: &str,
     mut last_block_header: SPVBlockHeader,
     storage: &dyn BlockHeaderStorageOps,
-    retarget_header: &SPVBlockHeader,
 ) -> Result<BlockHeaderBits, NextBlockBitsError> {
     if last_block_header.height == 0 {
         return Ok(BlockHeaderBits::Compact(MAX_BITS_BTC.into()));
@@ -141,7 +125,7 @@
 
     if is_retarget_height(height) {
         last_block_header.height = height;
-        btc_retarget_bits(coin, last_block_header, storage, retarget_header).await
+        btc_retarget_bits(coin, last_block_header, storage).await
     } else {
         Ok(last_block_bits)
     }
@@ -152,7 +136,6 @@
     current_block_timestamp: u32,
     mut last_block_header: SPVBlockHeader,
     storage: &dyn BlockHeaderStorageOps,
-    retarget_header: &SPVBlockHeader,
 ) -> Result<BlockHeaderBits, NextBlockBitsError> {
     let max_bits = BlockHeaderBits::Compact(MAX_BITS_BTC.into());
     if last_block_header.height == 0 {
@@ -165,7 +148,7 @@
 
     if is_retarget_height(height) {
         last_block_header.height = height;
-        btc_retarget_bits(coin, last_block_header, storage, retarget_header).await
+        btc_retarget_bits(coin, last_block_header, storage).await
     } else if current_block_timestamp > max_time_gap {
         Ok(max_bits)
     } else if last_block_bits != max_bits {
Index: mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs b/mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs
--- a/mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs	(revision d380e5c32a4364a98e9e0b9b173cfebc3fb5025a)
+++ b/mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs	(date 1674558847481)
@@ -370,16 +370,8 @@
             }
 
             if let Some(algorithm) = &params.difficulty_algorithm {
-                let retarget_header = &conf.block_header;
-                let next_block_bits = next_block_bits(
-                    coin,
-                    header.time,
-                    previous_header.clone(),
-                    storage,
-                    algorithm,
-                    retarget_header,
-                )
-                .await?;
+                let next_block_bits =
+                    next_block_bits(coin, header.time, previous_header.clone(), storage, algorithm).await?;
 
                 if !params.constant_difficulty && params.difficulty_check && current_block_bits != next_block_bits {
                     return Err(SPVError::InsufficientWork);
```

Reply:
ok thanks